File: examples/EmojiFun/bindings/dart/test/example_test.dart
import 'package:cricut_emojifun/emojifun.dart';
import 'package:test/test.dart';

void main() {
  setUp(() {
    final _ = ensureLoaded;
  });

  group('EmojiFunTests', () {
    test('hello', () {
        expect(EmojiExplorer.hello, equals("👋🌎"));
    });

    test('construction', () {
        var explorer = EmojiExplorer.create(128);
        expect(explorer.known.length, equals(128));
        expect(explorer.known, equals(Set.from([
            "⌚", "⌛", "⏩", "⏪", "⏫", "⏬", "⏰", "⏳", "◽", "◾", "☔", "☕", "♈", "♉", "♊", "♋",
            "♌", "♍", "♎", "♏", "♐", "♑", "♒", "♓", "♿", "⚓", "⚡", "⚪", "⚫", "⚽", "⚾", "⛄",
            "⛅", "⛎", "⛔", "⛪", "⛲", "⛳", "⛵", "⛺", "⛽", "✅", "✊", "✋", "✨", "❌", "❎", "❓",
            "❔", "❕", "❗", "➕", "➖", "➗", "➰", "➿", "⬛", "⬜", "⭐", "⭕", "🀄", "🃏", "🆎", "🆑",
            "🆒", "🆓", "🆔", "🆕", "🆖", "🆗", "🆘", "🆙", "🆚", "🇦", "🇧", "🇨", "🇩", "🇪", "🇫", "🇬",
            "🇭", "🇮", "🇯", "🇰", "🇱", "🇲", "🇳", "🇴", "🇵", "🇶", "🇷", "🇸", "🇹", "🇺", "🇻", "🇼",
            "🇽", "🇾", "🇿", "🈁", "🈚", "🈯", "🈲", "🈳", "🈴", "🈵", "🈶", "🈸", "🈹", "🈺", "🉐", "🉑",
            "🌀", "🌁", "🌂", "🌃", "🌄", "🌅", "🌆", "🌇", "🌈", "🌉", "🌊", "🌋", "🌌", "🌍", "🌎", "🌏",
        ])));

        expect(EmojiExplorer.create(0).known.length, equals(0));
        expect(EmojiExplorer.create(-1).known.length, equals(0));
        expect(EmojiExplorer.create(1212 + 50).known.length, greaterThan(1212));
    });

    test('random', () {
        var explorer = EmojiExplorer.create(128);
        for (var i = 0; i < 100; i++) {
            expect(explorer.random().length, greaterThan(0));
        }

        expect(EmojiExplorer.create(0).random(), equals(EmojiExplorer.hello));
    });

    test('randomUnique', () {
        const count = 128;
        var explorer = EmojiExplorer.create(count);

        var reported = <String>[];
        for (var i = 0; i < explorer.known.length; i++) {
            var emoji = explorer.randomUnique();
            if (emoji != null) {
                reported.add(emoji);
            }
        }

        expect(reported.length, equals(count));
        expect(Set.from(reported), equals(explorer.known));
    });

    test('enumerateKnown', () {
        const count = 128;
        var explorer = EmojiExplorer.create(count);

        var reported = <String>[];
        explorer.enumerateKnown((String s) {
            reported.add(s);
            return true;
        });

        expect(reported.length, equals(count));
        expect(Set.from(reported), equals(explorer.known));

        var earlyExitReportedCount = 0;
        explorer.enumerateKnown((String s) {
            earlyExitReportedCount += 1;
            return earlyExitReportedCount < 50;
        });
        expect(earlyExitReportedCount, equals(50));
    });
  });
}

File: Examples/EmojiFun/bindings/ts/tests/example.test.ts
import { EmojiFun } from 'EmojiFun';

test('hello', () => {
    expect(EmojiFun.EmojiExplorer.hello).toBe("👋🌎");
});

test('construction', () => {
    const explorer = EmojiFun.EmojiExplorer.create(128);
    expect(explorer.known.size).toBe(128);
    expect(Array.from(explorer.known.values()).sort()).toEqual([
        "⌚", "⌛", "⏩", "⏪", "⏫", "⏬", "⏰", "⏳", "◽", "◾", "☔", "☕", "♈", "♉", "♊", "♋",
        "♌", "♍", "♎", "♏", "♐", "♑", "♒", "♓", "♿", "⚓", "⚡", "⚪", "⚫", "⚽", "⚾", "⛄",
        "⛅", "⛎", "⛔", "⛪", "⛲", "⛳", "⛵", "⛺", "⛽", "✅", "✊", "✋", "✨", "❌", "❎", "❓",
        "❔", "❕", "❗", "➕", "➖", "➗", "➰", "➿", "⬛", "⬜", "⭐", "⭕", "🀄", "🃏", "🆎", "🆑",
        "🆒", "🆓", "🆔", "🆕", "🆖", "🆗", "🆘", "🆙", "🆚", "🇦", "🇧", "🇨", "🇩", "🇪", "🇫", "🇬",
        "🇭", "🇮", "🇯", "🇰", "🇱", "🇲", "🇳", "🇴", "🇵", "🇶", "🇷", "🇸", "🇹", "🇺", "🇻", "🇼",
        "🇽", "🇾", "🇿", "🈁", "🈚", "🈯", "🈲", "🈳", "🈴", "🈵", "🈶", "🈸", "🈹", "🈺", "🉐", "🉑",
        "🌀", "🌁", "🌂", "🌃", "🌄", "🌅", "🌆", "🌇", "🌈", "🌉", "🌊", "🌋", "🌌", "🌍", "🌎", "🌏",
    ].sort());

    expect(EmojiFun.EmojiExplorer.create(0).known.size).toBe(0);
    expect(EmojiFun.EmojiExplorer.create(-1).known.size).toBe(0);
    expect(EmojiFun.EmojiExplorer.create(1431 + 50).known.size).toBeGreaterThanOrEqual(1431);
});

test('random', () => {
    const explorer = EmojiFun.EmojiExplorer.create(128);
    for (let i = 0; i < 100; i++) {
        expect(explorer.random().length).toBeGreaterThan(0);
    }

    expect(EmojiFun.EmojiExplorer.create(0).random()).toBe(EmojiFun.EmojiExplorer.hello);
});

test('randomUnique', () => {
    const count = 128;
    const explorer = EmojiFun.EmojiExplorer.create(count);

    const reported: string[] = [];
    for (let i = 0; i < explorer.known.size; i++) {
        const emoji = explorer.randomUnique();
        if (emoji !== undefined) {
            reported.push(emoji);
        }
    }

    expect(reported.length).toBe(count);
    expect(reported.sort()).toEqual(Array.from(explorer.known.values()).sort());
});

test('enumerateKnown', () => {
    const count = 128;
    const explorer = EmojiFun.EmojiExplorer.create(count);

    const reported: string[] = [];
    explorer.enumerateKnown(s => {
        reported.push(s);
        return true;
    });

    expect(reported.length).toBe(count);
    expect(reported.sort()).toEqual(Array.from(explorer.known.values()).sort());

    let earlyExitReportedCount = 0;
    explorer.enumerateKnown(s => {
        earlyExitReportedCount += 1;
        return earlyExitReportedCount < 50;
    });
    expect(earlyExitReportedCount).toBe(50);
});
